libgdf/src/device_column.cuh

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright (c) 2018, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+#ifndef DEVICE_COLUMN_H
+#define DEVICE_COLUMN_H
+
+#include <memory>
+#include <gdf/gdf.h>
+
+// forward declaration
+struct DeviceColumn;
+
+// Alias for a unique pointer to a device allocated pointer which
+// points to a DeviceColumn constucted on the device. Custom deleter
+// calls a kernel which deletes the object constructed on the device
+// and frees the device allocation for the DeviceColumn*
+using DeviceColumnPointer = std::unique_ptr<DeviceColumn*, 
+                                            std::function<void(DeviceColumn**)>>;
+
+struct DeviceColumn
+{
+  __host__
+  static DeviceColumnPointer make_device_column(gdf_column col);
```

Comment:
Is there some way to do this with a copy constructor rather than a factory member function?

Reply:
Also, should `make_device_column` perhaps be defined outside of the `DeviceColumn` class, rather than as a static member function?  `DeviceColumn::make_device_column()` is a bit repetitive.

Reply:
Factory functions (or classes) are pretty much the standard practice for creating a new instance of one of many Derived classes, where the specified derived class depends on some user input (in this case the type of the gdf_column).

Furthermore, it's the standard practice of creating factory functions to make them a static member of the base class (though I'm having trouble explaining exactly _why_ is the case). I'll change the name to `DeviceColumn::create()` so it's shorter for you :)

Reply:
OK. It's good to learn the standard practice, thanks. FWIW, I do prefer "make_x" naming, since it also seems to be standard practice, e.g. `std::make_pair()`.